src/bin/pg_autoctl/service_keeper.c

```diff
@@ -511,6 +558,31 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
 			}
 		}
 
+		/*
+		 * When we receive SIGTERM, we enable maintenance on the monitor, and
+		 * we record this action in the Keeper State file. At startup, we then
+		 * disable maintenance again. We might have to retry disabling
+		 * maintenance should it fail.
+		 */
+		if (shutdownInMaintenance && !disabledMaintenanceAtStartup)
+		{
+			int nodeId = keeper->state.current_node_id;
+			bool mayRetry = false;
+
+			log_info("pg_autoctl node was put in maintenance as part of "
```

Comment:
When after shutting down pg_autoctl, you run `pg_autoctl disable maintenance` manually, then we get in an infinite loop here:
```
12:59:22 22037 INFO  pg_autoctl node was put in maintenance as part of shutting down, now disabling maintenance
12:59:22 22037 ERROR Monitor ERROR:  cannot stop maintenance when current state for node 1 "node1" (localhost:5501) is not "maintenance"
12:59:22 22037 ERROR Monitor DETAIL:  Current reported state is "maintenance" and assigned state is "catchingup"
12:59:22 22037 ERROR Failed to stop_maintenance of node 1 from the monitor
```

Reply:
That's the new code that handles this part now. Thanks to fixing the case when we're in prepare_maintenance at startup and want to finish the FSM transition to maintenance and immediately disable maintenance, I believe this comment is fixed too. At least I'm not sure how to reproduce.

``` C
		/*
		 * When we receive SIGTERM, we enable maintenance on the monitor, and
		 * we record this action in the Keeper State file. At startup, we then
		 * disable maintenance again. We might have to retry disabling
		 * maintenance should it fail.
		 *
		 * The shutdown sequence might also have been interrupted in
		 * wait_maintenance state, in which case we want to first finish the
		 * FSM transition to maintenance, and only later disable maintenance.
		 */
		if (shutdownInMaintenance &&
			!disabledMaintenanceAtStartup &&
			keeperState->assigned_role == MAINTENANCE_STATE &&
			keeperState->current_role == keeperState->assigned_role &&
			!needStateChange)
		{
			int nodeId = keeper->state.current_node_id;
			bool mayRetry = false;

			log_info("pg_autoctl node was put in maintenance as part of "
					 "shutting down, now disabling maintenance");
```

Reply:
if (!disabledMaintenannceAtStartup && (
        keeperState->assigned_role != MAINTENANCE_STATE ||
        keeperState->assigned_role != PREPARE_MAINTENANCE_STATE ||
        keeperState->assigned_role != WAIT_MAINTENANCE_STATE ||
)) { 
    disabledMaintenannceAtStartup = true
}

Reply:
change !disabledMaintenanceAtStartup to shouldDisableMaintenanceAtStartup

Reply:
Done now, to cover for the following test case / scenario:

  1. pg_autoctl stop
  2. pg_autoctl disable maintenance
  3. pg_autoctl run
  4. pg_autoctl enable maintenance

Step 4 would do a secondary->maintenance->secondary without the fix outlined here.